### DISCUSSION TITLE
All settings should have a consistent prefix

### DISCUSSION THREAD

**Discussion by gerardroche** [NeoVintageous/NeoVintageous#927] [General]
# Refactor settings

- [ ] Rename prefix: `nv.`
- [ ] Rename plugin prefixes: `{prefix}{plugin_name}`
- :heavy_check_mark: means no change
- Actions are currently not set in stone

| Setting                                           | Default   | Type      | Action
| :------------------------------------------------ | :-------- | :-------- | :----------
| vintageous_auto_complete_exit_from_insert_mode    | true      | Boolean   |
| vintageous_auto_nohlsearch_on_normal_enter        | true      | Boolean   | :heavy_check_mark:
| vintageous_auto_switch_input_method               | false     | Boolean   |
| vintageous_auto_switch_input_method_default       | ""        | String    |
| vintageous_auto_switch_input_method_get_cmd       | ""        | String    |
| vintageous_auto_switch_input_method_set_cmd       | ""        | String    |
| vintageous_bell                                   | "blink"   | String    | Rename: visual_bell_style or visual_bell_type
| vintageous_bell_color_scheme                      | "dark"    | String    | Rename: visual_bell_color_scheme
| vintageous_clear_auto_indent_on_esc               | true      | Boolean   |
| vintageous_default_mode                           | ""        | String    | Change default: null
| vintageous_enable_abolish                         | true      | Boolean   | Rename: abolish_enable
| vintageous_enable_commentary                      | true      | Boolean   | Rename: commentary_enable
| vintageous_enable_multiple_cursors                | true      | Boolean   | Rename: multiple_cursors_enable
| vintageous_enable_sneak                           | false     | Boolean   | Rename: sneak_enable
| vintageous_enable_sublime                         | true      | Boolean   | Rename: sublime_enable
| vintageous_enable_surround                        | true      | Boolean   | Rename: surround_enable
| vintageous_enable_unimpaired                      | true      | Boolean   | Rename: unimpaired_enable
| vintageous_exit_when_quitting_last_window         | true      | Boolean or String | Maybe rename vintageous_exit_when_quitting_last_view
| vintageous_handle_keys                            | \{\}      | Dict      | :heavy_check_mark:
| vintageous_highlighted_yank                       | true      | Boolean   | :heavy_check_mark:
| vintageous_highlighted_yank_duration              | 1000      | Integer   | :heavy_check_mark:
| vintageous_highlighted_yank_style                 | "fill"    | String    | :heavy_check_mark:
| vintageous_i_escape_jj                            | false     | Boolean   | :heavy_check_mark:
| vintageous_i_escape_jk                            | false     | Boolean   | :heavy_check_mark:
| vintageous_i_escape_kj                            | false     | Boolean   | :heavy_check_mark:
| vintageous_lsp_save                               | false     | Boolean   | Maybe rename vintageous_save_lsp because there is vintageous_save_async
| vintageous_multi_cursor_exit_from_visual_mode     | false     | Boolean   |
| vintageous_reset_mode_when_switching_tabs         | true      | Boolean   |
| vintageous_save_async                             | false     | Boolean   |
| vintageous_search_cur_style                       | "fill"    | String    | Rename: search_current_style or search_style_current?
| vintageous_search_inc_style                       | "fill"    | String    | Rename: search_incremental_style or search_style_sncremental
| vintageous_search_occ_style                       | "fill"    | String    | Rename: search_occurrences_style or search_style_occurrences
| vintageous_shell_silent                           | false     | Boolean   | Not sure this is needed because we have the :silent command. This setting was probably made available before the ex command was added.
| vintageous_show_marks_in_gutter                   | true      | Boolean   | :heavy_check_mark:
| vintageous_sneak_use_ic_scs                       | 0         | Integer   | :heavy_check_mark:
| vintageous_source                                 | null      | String    | :heavy_check_mark: Can maybe be expanded to allow a list of sources.
| vintageous_terminal                               | null      | String    |
| vintageous_use_ctrl_keys                          | true      | Boolean   | :white_check_mark: Maybe enable_ctrl_keys. See also use_super_keys and use_sys_clipboard
| vintageous_use_super_keys                         | false     | Boolean   | :white_check_mark:
| vintageous_use_sys_clipboard                      | false     | Boolean   | Rename: use_system_clipboard

**Notes**

- Visual bells are disabled via `'belloff'` option, hence no Boolean vintageous_bell setting. The bell related settings are for configuring the bell.

**Changelog**

- Deprecated vintageous_exit_when_quiting_last_window; use vintageous_exit_when_quitting_last_window instead (done with backwards compatibility)
- Deprecated highlightedyank; use vintageous_highlighted_yank instead (done with backwards compatibility)
- Deprecated highlightedyank_duration; use vintageous_highlighted_yank_duration instead (done with backwards compatibility)
- Deprecated highlightedyank_style; use vintageous_highlighted_yank_style instead (done with backwards compatibility)
- Deprecated neovintageous_search_cur_style; use vintageous_search_cur_style instead (done with backwards compatibility)
- Deprecated neovintageous_search_inc_style; use vintageous_search_inc_style instead (done with backwards compatibility)
- Deprecated neovintageous_search_occ_style; use vintageous_search_occ_style instead (done with backwards compatibility)
- Changed vintageous_i_escape_jj default from null to false (no meaningful functional difference for end users)
- Changed vintageous_i_escape_jk default from null to false (no meaningful functional difference for end users)
- Changed vintageous_i_escape_kj default from null to false (no meaningful functional difference for end users)

---

1. What should the setting {prefix} be?

   - ~`neovintageous_`~
   - ~`neovintageous.`~
   - ~`nv_`~
   - `nv.` :heavy_check_mark:

   "nv." is the current choice.

2. Should a {plugin_prefix} be?

   - `{prefix}{plugin_name}_` :heavy_check_mark:
   - ~`{prefix}{plugin_name}.`~
   - ~No~

3. Should plugins consistently have an boolean enabled setting:

   - `{plugin_prefix}_enable`
   - `{plugin_prefix}`
   - ~`{plugin}_{plugin_name}_enable`~

---

I'm leaning towards the prefix `neovintageous.` with the full stop instead of underscore.

But for plugins I think maybe use the underscore:

`neovintageous.{plugin_name}_` **vs** `neovintageous.{plugin_name}.`

Examples

```
neovintageous.highlighted_yank_enable
neovintageous.highlighted_yank_duration
neovintageous.highlighted_yank_style
```

vs

```
neovintageous.highlighted_yank.enable
neovintageous.highlighted_yank.duration
neovintageous.highlighted_yank.style
```

However, when I look at the examples, the full stop one looks better. All other settings would use underscore.

So for example, take the `vintageous_lsp_save` setting. This is a plugin specific setting. If a full stop is used for plugins, it makes it a little more clear. Consider

```
neovintageous.lsp.save
```

vs

```
neovintageous.lsp_save
```

It seems more cohesive. The  actual setting name, "save", in this case could maybe have a more descriptive name, but the full stop helps separate the  idea of plugin settings, be it out-of-box or third party, from the base neovintageous settings.

What are your suggestions? Thoughts?

**Comment by sourcevault** [comment]
**What should settings be prefixed with?**

``nv`` is fine.


**Should all plugins be prefixed {prefix}{plugin_name}_?**

I do not see the value of the underscore.


**Should plugins be enabled via a boolean {prefix}{plugin_name}_enable ?**

simple boolean is fine, no need for `enable`.


﹋﹋﹋﹋﹋﹋﹋﹋
```
neovintageous.highlighted_yank_enable
neovintageous.highlighted_yank_duration
neovintageous.highlighted_yank_style
```

vs

```
neovintageous.highlighted_yank.enable
neovintageous.highlighted_yank.duration
neovintageous.highlighted_yank.style
```
﹏﹏﹏﹏﹏﹏﹏﹏

Alright, I think I understand where you are coming from.

If you programmed in Lua for a while you would grow to be more comfortable around the heavy use of underscore for nesting, and after a while you would see it as the same as a `.` (dot).

Your answer depends of the depth of your tree, if there are large number of downstream variables (`enable`, `duration``style`) then the dot makes more sense.

If you have just mostly a single variable then the underscore makes more sense.

For consistency the dot is better.

**Reply by gerardroche** [reply]
Yeah, `nv` is nice and short. Underscore as separator, except the main namespace e.g.  `nv.use_ctrl_keys`.

In nearly all my plugins I use this convention of using a dot to indicate the plugin namespace and then underscore as a separator. It just seems to make sense to me visually.

**Comment by eugenesvk** [comment]
plugins with this many settings should not pollute the main settings file, and this would neatly avoid the issue of prefixes altogether since the "prefix" becomes a separate file like for other plugins

And then to avoid further unwieldy repetitions in names like
```
highlighted_yank_enable
highlighted_yank_duration
highlighted_yank_style
```
you can use nesting

**Reply by sourcevault** [reply]
> plugins with this many settings should not pollute the main settings file, and this would neatly avoid the issue of prefixes altogether since the "prefix" becomes a separate file like for other plugins.

Many complicated programs like Autodesk Maya have hundreds of configuration options at the top level and look exactly like what you have described.

Having a single configuration file is a lot less complex than dealing with multiple ones.

Sublime's own `Preferences.sublime-settings` and `Defaults.sublime-keymap` file has close to a thousand options.

Neovint for now has not reached that stage and when it too has +1000 options in a single file then it would make sense to split it.

Having everything in a single file also ensures you can you regex search quite easily.

**Reply by eugenesvk** [reply]
> Many complicated programs like Autodesk Maya have hundreds of configuration options at the top level and look exactly like what you have described.

There are plenty of poorly designed apps out there, so?

> Having a single configuration file is a lot less complex than dealing with multiple ones.

And in other ways it's more complex: harder to read (verbose repetitive key names), harder to find

> Sublime's own `Preferences.sublime-settings` and `Defaults.sublime-keymap` file has close to a thousand options.

189 preferences and 333 keys, they just have this vertically wasteful format
```json
	// Font options for theme text. Valid options are the same as
	// "font_options", except "no_bold" and "no_italic" are not supported
	"theme_font_options": [],
```
instead of 
```
	"theme_font_options": [], // Valid options are the same as "font_options", except "no_bold" and "no_italic"
```

But then again, so what? Their config format is poorly structured and also very verbose with field names taking more space than the key content. Why would you emulate that?

> Neovint for now has not reached that stage and when it too has +1000 options in a single file then it would make sense to split it.

Or you could choose scalable design now and not waste users' time later. Also, NV is adding on top of whatever part of that 1000 the user has changed

> Having everything in a single file also ensures you can you regex search quite easily

You get the same benefit via while regex-searching your config project. But then you'd also be able to collapse irrelevant groups
But if everything is one file, you can't easily exclude irrelevant options

**Reply by gerardroche** [reply]
I prefer having my settings in one file. I usually on tweak at most a few plugin settings. Even when Sublime stripped comments from the settings file, I still preferred it. I just sorted my settings and well prefixed plugin settings grouped together nicely. Now that Sublime keeps the comments, I group my plugin settings into sorted sections.

**Reply by gerardroche** [reply]
Tip: Sort sections with `<leader>ss`: `nnoremap <leader>ss vip:sort u<CR>`

**Reply by eugenesvk** [reply]
> I usually on tweak at most a few plugin settings

so what about those users who tweak more than a few?

> well prefixed plugin settings grouped together nicely

Semantic sorting is better than alphabetical. For example, all plugin settings could be in a separate section from "core editing" config

> when Sublime stripped comments from the settings file

that was one of my worst initial experiences with Sublime, such a disrespect to user's effort in config organization

**Reply by sourcevault** [reply]
How would you export multiple setting files as `.pdf` ?

I usually keep a pdf of the setting file and use `ctrl+f` to browse it on a seperate window, that is a lot more complicated once you have multiple files.

Also that makes it easy to find key conflicts with other plugins.

There is a lot more to this question that what meets the eyes.

**Reply by eugenesvk** [reply]
> How would you export multiple setting files as `.pdf` ?

- How do you do it know when a lot of plugins have separate settings files?
- Why would you ever use that awful format and lose all the power of Sublime in the process?

> I usually keep a pdf of the setting file and use `ctrl+f` to browse it on a seperate window,

For example, how do you regex-search in a PDF?

>   that is a lot more complicated once you have multiple files.

No it's not, I've already replied to you about this point in the first comment

> You get the same benefit via while regex-searching your config project. But then you'd also be able to collapse irrelevant groups
But if everything is one file, you can't easily exclude irrelevant options

So just have a Sublime project with your config files, and search within all files in a project, that's much better than any pdf, and you can also have it opened in a separate window

> Also that makes it easy to find key conflicts with other plugins.

What key conflicts??? There are no keybinds in a config, that's what keymaps are for

> There is a lot more to this question that what meets the eyes.

Not really, this is a very common practices other plugins have been employing for ages
